### PR TITLE
[AIR] Ensure GBDT PGFs match passed ScalingConfig

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Depends on https://github.com/ray-project/ray/pull/30304

This PR fixes an issue with the special case from `GBDTTrainer`, where the `ScalingConfig` would be converted to `xgboost_ray.RayParams` first before being converted to a PGF for purposes of determining trial resources. This however results in dropping of strategy, args and kwargs from `ScalingConfig`, as `RayParams` doesn't support them. This PR introduces backwards compatible logic to deal with this issue. The changes introduced here will be upstreamed to `xgboost_ray`. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by amogkam — COMMENTED** [review]
thanks @Yard1 lgtm. can we add a small test?

**Comment by Yard1** [comment]
This is tested by ensuring that the strategy is not dropped from PGF in the resource changing test, but I can add another one

**Comment by Yard1** [comment]
@amogkam added, PTAL

**Review by amogkam — APPROVED** [review]
Thanks!

File: python/ray/train/gbdt_trainer.py [hunk 1]
```diff
@@ -59,9 +60,38 @@ def _convert_scaling_config_to_ray_params(
...
         }
     )
-    ray_params = ray_params_cls(
+
+    # This should be upstreamed to xgboost_ray,
+    # but also left here for backwards compatibility.
+    if not hasattr(ray_params_cls, "placement_strategy"):
+
+        class RayParamsFromScalingConfig(ray_params_cls):
+            placement_strategy: Optional[str] = None
```
Comment (krfricke) [inline python/ray/train/gbdt_trainer.py:69]:
```suggestion
            placement_strategy: str = "PACK"
```
it currently defaults to PACK in xgboost-ray, so this should probably be reflected here

File: python/ray/train/gbdt_trainer.py [hunk 2]
```diff
@@ -59,9 +60,38 @@ def _convert_scaling_config_to_ray_params(
...
+            pgf_args: Optional[Tuple[Any]] = None
+            pgf_kwargs: Optional[Dict[str, Any]] = None
```
Comment (krfricke) [inline python/ray/train/gbdt_trainer.py:71]: Can we just have a single option `placement_options` that contains kwargs to pass to PGF? If we want to upstream this, we should try to keep the `RayParams` dataclass clean as it is also used without Tune

File: python/ray/train/gbdt_trainer.py [hunk 3]
```diff
@@ -59,9 +60,38 @@ def _convert_scaling_config_to_ray_params(
...
+    else:
+        ray_params_cls_extended = ray_params_cls
+
+    ray_params = ray_params_cls_extended(
         **ray_params_kwargs,
     )
+    scaling_config_pgf = scaling_config.as_placement_group_factory()
+    ray_params.placement_strategy = scaling_config.placement_strategy
+    ray_params.pgf_args = scaling_config_pgf._args
+    ray_params.pgf_kwargs = scaling_config_pgf._kwargs
```
Comment (krfricke) [inline python/ray/train/gbdt_trainer.py:94]: IMO it's not great to convert the scaling config into a PGF here and then cherry-pick specific attributes to pass into the PGF.

I understand that this ensures other arguments (e.g. cpu fraction per node) are automatically caught, but maybe for now we can just build the `placement_options` above manually. Feel free to add a comment into `ScalingConfig.as_placement_group` to keep in sync with GBDT RayParams

**Review by krfricke — CHANGES_REQUESTED** [review]
Good catch, but a couple of small requests

**Review by krfricke — APPROVED** [review]
Thanks! Can you open a PR upstream at XGBoost-Ray for these changes?

**Pull request merged**